libcxx/include/__iterator/prev.h

```diff
@@ -35,6 +39,13 @@ prev(_InputIter __x, typename iterator_traits<_InputIter>::difference_type __n =
   return __x;
 }
 
+template <class _BidirectionalIterator,
+          __enable_if_t<__has_bidirectional_iterator_category<_BidirectionalIterator>::value, int> = 0>
```

Comment:
About `enable_if` vs `static_assert`. Upon re-consideration, I think the right thing to do here would be a `static_assert`. Indeed, we have the runtime assertion `_LIBCPP_ASSERT_PEDANTIC(__n <= 0 || __has_bidirectional_iterator_category<_InputIter>::value)` above, and one way to think about this patch is that we're only "inlining" the knowledge that `n == 1` in this overload. That gives us:

```
_LIBCPP_ASSERT_PEDANTIC(false || __has_bidirectional_iterator_category<_InputIter>::value)
```

but then we might as well check that at compile-time since there is no runtime-dependent value anymore. So we get

```
static_assert(__has_bidirectional_iterator_category<_InputIter>::value)
```

for this overload.

That also means we should switch to a `.verify.cpp` test instead.

Reply:
The intent of my comment is that we could replace the runtime assertion above by:

```cpp
if constexpr (__builtin_constant_p(__n) && __n > 0) {
  static_assert(__has_bidirectional_iterator_category<_InputIter>::value, "this makes no sense");
} else {
  _LIBCPP_ASSERT_PEDANTIC(__n <= 0 || __has_bidirectional_iterator_category<_InputIter>::value);
}
```

Essentially, we're making sure to provide a compile-time error when we can. And using a separate overload is just one way of achieving that (although I guess the `__builtin_constant_p` approach is even better). One problem is that this doesn't work pre-C++17.

Reply:
sure. stick to two overloads approach with both enable if on the "input_iterator_tag" and added `static_assert` on the second overload